src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -1997,18 +2033,49 @@ public void shutdown() {
         combinedCache.shutdownNow();
       }
       Thread.currentThread().interrupt();
+      return PendingUploads.EMPTY;
     }
 
-    // Waits for all background tasks to finish and interrupts them if there is another interrupt.
-    backgroundTaskExecutor.close();
-
-    // Release the cache only after background tasks are done as they might be using it.
-    if (combinedCache != null) {
-      combinedCache.release();
-    }
+    if (cacheAsync != RemoteCacheAsync.NOWAIT) {
+      // Existing behavior - block until done
+      backgroundTaskExecutor.close();
+      if (combinedCache != null) {
+        combinedCache.release();
+      }
+      if (remoteExecutor != null) {
+        remoteExecutor.close();
+      }
+      return PendingUploads.EMPTY;
+    } else {
+      // NOWAIT mode - return handle for caller to wait/cancel later
+      return new PendingUploads() {
+        @Override
+        public void awaitTermination() throws InterruptedException {
+          // Wait for all background tasks to complete
+          backgroundTaskExecutor.shutdown();
+          while (!backgroundTaskExecutor.awaitTermination(1, java.util.concurrent.TimeUnit.DAYS)) {
+            // Keep waiting
+          }
+          // Release resources after uploads complete
+          if (combinedCache != null) {
+            combinedCache.release();
+          }
+          if (remoteExecutor != null) {
+            remoteExecutor.close();
+          }
+        }
 
-    if (remoteExecutor != null) {
-      remoteExecutor.close();
+        @Override
+        public void cancel() {
+          backgroundTaskExecutor.shutdownNow();
```

Comment:
```suggestion
          backgroundTaskExecutor.shutdownNow();
          backgroundTaskExecutor.close()
```
to await the interrupted tasks

Reply:
Done.